segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+
+            long lastSnapshotOffset = getLastSnapshotOffset();
+            int readLength = (int) Math.min(Integer.MAX_VALUE, this.attributeSegment.get().getLength() - lastSnapshotOffset);
+            CompletableFuture<AttributeCollection> r = new CompletableFuture<>();
+            if (readLength == 0) {
+                // Nothing to read.
+                r.complete(new AttributeCollection());
+            } else {
+                AsyncReadResultProcessor.process(
+                        StreamSegmentStorageReader.read(this.attributeSegment.get().handle, lastSnapshotOffset, readLength, this.config.getReadBlockSize(), this.storage),
+                        new AttributeSegmentReader(r, timeout),
+                        this.executor);
+            }
+            return r;
+        }, this.executor);
+        return Futures.exceptionallyCompose(result,
+                ex -> {
+                    ex = Exceptions.unwrap(ex);
+                    if (ex instanceof IOException || ex instanceof StreamSegmentTruncatedException) {
+                        ex = new DataCorruptionException(
+                                String.format("Unable to parse AttributeSegment. LastSnapshot = (Offset=%d, Length=%d), Known Segment Length = %d.",
+                                        getLastSnapshotOffset(), getLastSnapshotLength(), this.attributeSegment.get().getLength()), ex);
+                    }
+                    return Futures.failedFuture(ex);
+                });
+    }
+
+    /**
+     * Creates a new Snapshot, writes it to the AttributeSegment and updates the main Segment's attributes with its location.
+     *
+     * @param newAttributes Any new Attributes to include in this Snapshot that would not already be part of it.
+     * @param mustComplete  Whether the update of the main Segment's attributes must execute successfully in order for this
+     *                      operation to be considered complete. For auto-generated Snapshots, since the Snapshot itself is
+     *                      already written to the AttributeSegment, it is not mandatory for the OperationLog Add &
+     *                      Attribute Segment Truncation operations to complete, as there will be no data loss. However, for
+     *                      Snapshots generated as part of Seals, all of those must complete in order for the operation to
+     *                      be considered successful.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation has completed successfully.
+     */
+    private CompletableFuture<Void> createSnapshot(AttributeCollection newAttributes, boolean mustComplete, Duration timeout) {
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+
+        // The serialization may be invoked multiple times, based on whether the appendConditionally() requires a retry.
+        Supplier<CompletableFuture<ArrayView>> s = () ->
```

Comment:
A clarification: here we are generating a new snapshot by merging the attributes in the last snapshot with the new attributes. Does it mean that we never really get rid of attributes, unless they are explicitly deleted?

 I'm asking because of the case of writer ids, in which we can have a writer crashing and never coming back. We will never get rid of the corresponding writer id, and worse, we will keep accumulating such ids.

Reply:
The only way to get rid of an attribute is to set its value to `Attributes.NULL_ATTRIBUTE_VALUE`. Otherwise I cannot remove it. 

The problem with writer ids never getting cleaned up is beyond the scope of this PR and should be addressed in the Client/Wire Protocol.

Reply:
It is fine to declare it outside scope, but I'm not sure we can address it in the client/wire protocol simply because a crashed client won't remember its writer id. We might need to expire.

Would you mind creating an issue so that we can discuss and converge on a plan of action?

Reply:
#2509